### PR TITLE
Improve proto generation

### PR DESCRIPTION
**Pull request by r2k1**
/kind cleanup

- generate code locally using standard protoc tool instead of generating it remotely on buf servers
- run tools in docker instead of executing local binaries, use consistent buf/protobuf tools on all dev machines
- update README

Note: I didn't test it on amd64. May need to adjust dockerfile if I put an incorrect value there.

### CODE REVIEW

File: aks-node-controller/pkg/gen/aksnodeconfig/v1/tls_bootstrapping_config.pb.go
Comment (timmy-wright) [inline aks-node-controller/pkg/gen/aksnodeconfig/v1/tls_bootstrapping_config.pb.go:1]: This file should be removed - it's based on a file that was renamed/deleted.

Comment (r2k1) [reply aks-node-controller/pkg/gen/aksnodeconfig/v1/tls_bootstrapping_config.pb.go:1]: updated makefile to delete such files

**Review by timmy-wright — APPROVED** [review]

**Review by lilypan26 — APPROVED** [review]

File: aks-node-controller/README.md [hunk 1]
```diff
@@ -96,19 +101,40 @@ write_files:
...
+    participant Client as Client
+    participant ARM as Azure Resource Manager (ARM)
+    participant VM as Virtual Machine (VM)
+
+    Client->>ARM: Request to create VM<br/>with CustomData & CSE
+    ARM->>VM: Deploy config.json<br/>(CustomData)
+    note over VM: cloud-init handles<br/>config.json deployment
+
+    note over VM: cloud-init completes processing
+    note over VM: Start aks-node-controller.service (systemd service)<br/> afer cloud-init
```
Comment (lilypan26) [inline aks-node-controller/README.md:122]: nit: typo `after`.

Thanks for reworking the diagram, much easier to follow.

File: aks-node-controller/README.md [hunk 2]
```diff
@@ -4,7 +4,7 @@ This directory contains files related to AKS Node Controller go binary.
 
 ## Overview
 
-AKS Node Controller is a go binary that is responsible for bootstrapping AKS nodes. The controller expects a predefined contract from the client of type [`AKSNodeConfig`](https://github.com/Azure/AgentBaker/tree/dev/pkg/proto/aksnodeconfig/v1) containing the bootstrap configuration. The controller has two primary functions: 1. Parse the bootstrap config and kickstart bootstrapping and 2. Monitor the completion status. 
+AKS Node Controller is a go binary that is responsible for bootstrapping AKS nodes. The controller expects a predefined contract from the client of type [`AKSNodeConfig`](https://github.com/Azure/AgentBaker/tree/dev/pkg/proto/aksnodeconfig/v1) containing the bootstrap configuration. The controller has two primary functions: 1. Parse the bootstrap config and kickstart bootstrapping and 2. Monitor the completion status.
```
Comment (lilypan26) [inline aks-node-controller/README.md:7]: Can you update the paragraph below to remove the reference to `GetNodeBootstrapping` now that it is outdated?

File: aks-node-controller/Makefile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# Run buf in docker, mounting the full repo into the container
+# Emulate running "buf" in the current directory
+BUF = docker run --volume "$(CURDIR)/../:$(CURDIR)/../" --workdir $(CURDIR) bufbuild/buf:1.47.2
+
+.PHONY: proto-generate
```
Comment (Devinwong) [inline aks-node-controller/Makefile:5]: Do we need to manually run `make proto-generate` before pushing the commit?

File: aks-node-controller/README.md [hunk 4]
```diff
@@ -96,19 +101,40 @@ write_files:
...
+2. CSE: Script used to poll bootstrap status and return exit status once complete.
 
 CSE script: `/opt/azure/containers/aks-node-controller provision-wait`
 
 
 #### Provisioning flow diagram:
 
-![provisionFlowDiagram](../.github/images/scriptlessBootstrapFlow.png)
+```mermaid
+sequenceDiagram
```
Comment (Devinwong) [inline aks-node-controller/README.md:112]: Great we can 'write' a diagram!

File: aks-node-controller/Makefile (hunk 3 above)
Comment (r2k1) [reply aks-node-controller/Makefile:5]: yes, CI may complain about it. But I didn't test it.

File: aks-node-controller/Makefile [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+	@($(BUF) format -w)
+	rm -rf pkg/gen/aksnodeconfig/v1
+	docker build --platform $(shell uname -m) -t protoc-docker - < protoc.Dockerfile
+	docker run --rm -v $(shell pwd):/$(shell pwd) --workdir=$(shell pwd) protoc-docker protoc --go_opt=module=github.com/Azure/agentbaker/aks-node-controller --go_out=./  --proto_path=proto  $(shell find proto/aksnodeconfig/v1 -name '*.proto')
+	$(MAKE) proto-lint
+
+.PHONY: proto-lint
+proto-lint:
+	@($(BUF) lint)
+	@($(BUF) breaking --against '../.git#branch=dev,subdir=aks-node-controller') # TODO: change to master
```
[the remark this answers is not in the document]
Comment (r2k1) [reply aks-node-controller/Makefile:16]: couldn't find anything in the buf docs, I don't think it's supported. But CI behaves differently, it checks target branch for PRs.

**Pull request merged**